File: javascript/matmul.js
#!/usr/bin/env k8

function matgen(n) {
	let a = [], tmp = 1. / n / n;
	for (let i = 0; i < n; ++i) {
		a[i] = new Float64Array(n);
		for (let j = 0; j < n; ++j)
			a[i][j] = tmp * (i - j) * (i + j);
	}
	return a;
}

function matmul(a, b) {
	if (a[0].length != b.length) return;
	let c = [];
	for (let i = 0; i < a.length; ++i)
		c[i] = new Float64Array(b[0].length);
	for (let i = 0; i < a.length; ++i)
		for (let k = 0; k < b.length; ++k) {
			const t = a[i][k], l = b[0].length;
			for (let j = 0; j < l; ++j)
				c[i][j] += t * b[k][j];
		}
	return c;
}

var ccc = {
	print: typeof print == "function"? print : console.log,
	argv: typeof k8_version == "function"? arguments.slice(0) : typeof Deno == "object"? Deno.args.slice(0) : typeof Bun == "function"? Bun.argv.slice(2) : process.argv.splice(2)
};

let n = ccc.argv.length > 0? parseInt(ccc.argv[0]) : 1000;
const a = matgen(n), b = matgen(n);
const c = matmul(a, b);
ccc.print(c[n>>1][n>>1]);

File: python/matmul.py
#!/usr/bin/env python

import sys, array

# Writen by Attractive Chaos; distributed under the MIT license
# Modified by brentp: using array instead of list

def matmul(a, b):
	if len(a[0]) != len(b): raise
	ra, rb, rb0 = list(range(len(a))), list(range(len(b))), list(range(len(b[0])))
	d = [array.array('d', [0 for j in rb0]) for i in ra]
	for i in ra:
		for k in rb:
			aik, di, bk = a[i][k], d[i], b[k] # faster 
			for j in rb0:
				di[j] += aik * bk[j]
	return d

def main():
	n = 1500
	if (len(sys.argv) > 1): n = int(sys.argv[1])
	tmp = 1. / n / n
	a = [[tmp * (i - j) * (i + j) for j in range(n)] for i in range(n)]
	b = [[tmp * (i - j) * (i + j) for j in range(n)] for i in range(n)]
	d = matmul(a, b)
	print(d[n//2][n//2])

if __name__ == '__main__': main()
